postgres-persistence/src/main/java/com/netflix/conductor/postgres/util/PostgresQueueListener.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright 2024 Conductor Authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package com.netflix.conductor.postgres.util;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Iterator;
+
+import javax.sql.DataSource;
+
+import org.postgresql.PGConnection;
+import org.postgresql.PGNotification;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.conductor.core.exception.NonTransientException;
+import com.netflix.conductor.postgres.config.PostgresProperties;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class PostgresQueueListener {
+
+    private PGConnection pgconn;
+
+    private Connection conn;
+
+    private DataSource dataSource;
+
+    private HashMap<String, QueueStats> queues;
+
+    private boolean connected = false;
+
+    private boolean connecting = false;
+
+    private long lastNotificationTime = 0;
+
+    private Integer stalePeriod;
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    public PostgresQueueListener(DataSource dataSource, PostgresProperties properties) {
+        logger.info("Using experimental PostgresQueueListener");
+        this.dataSource = dataSource;
+        this.stalePeriod = properties.getExperimentalQueueNotifyStalePeriod();
+        connect();
+    }
+
+    public boolean hasMessagesReady(String queueName) {
+        checkUpToDate();
+        handleNotifications();
+        if (notificationIsStale() || !connected) {
+            connect();
+            return true;
+        }
+
+        QueueStats queueStats = queues.get(queueName);
```

Comment:
Thinking out loud while learning :grin:. There's multiple methods where a queue.get() is done, which may be done concurrently with a full overwrite of the queues member to a new map. If we don't synchronise this, hasMessagesReady and getSize methods may read out-of-date data. Is this a problem? 

I'm thinking no, since the queue is supposed to be polled periodically by the workers anyways, the off chance of missing a last update to this is better than the overhead of synchronization.

Reply:
Yeah, my thinking was that the data will move so quickly that this shouldn't be an issue. In a way, client 1 will get the latest data at the time they call the function and if client 2 changes the data while they are accessing the old data then that's fine.

Reply:
Entirely reasonable, I don't think the synchronization overhead is worth it.